### DISCUSSION TITLE
Use of MERRA2 data for NoahMp modeling

### DISCUSSION THREAD

**Discussion by Madhuka1990** [NCAR/noahmp#144] [General]
I am new to NoahMP modeling and trying to run NoahMP with MERRA2 data in 1km resolution. I regridded the data into geo_em file (means 1km) by using ESMF regrding script. 
I'm wondering whether I need to run create_forcing.exe again or just continue to run the model.

**Comment by cenlinhe** [comment]
If you already have the forcing data from MERRA2 regridded to geo_em file grids, then you can just go and run the model. Also, the current create_forcing.exe cannot handle MERRA2 data.

**Reply by Madhuka1990** [reply]
Thanks for quick reply.

**Reply by Madhuka1990** [reply]
If I don't run the create_forcing.exe, how do I create "HRLDAS_SETUP_FILE"?

Could you please give some idea

**Comment by Madhuka1990** [comment]
If I don't run the create_forcing.exe, how do I create "HRLDAS_SETUP_FILE"?

Could you please give some idea

**Reply by cenlinhe** [reply]
The most straightforward way to create a HRLDAS_SETUP_FILE is to use ERA5 or GLDAS or NLDAS data (just 1 or 2 time steps data) to run create_forcing.exe to generate the HRLDAS_SETUP_FILE. If your simulation includes spin up, then you do not need to replace the variable values in this HRLDAS_SETUP_FILE. If your spin up is short or no spin up, you can use MERRA2 variable values and regridded to the geo_em grid, and then replace the values in the generated HRLDAS_SETUP_FILE from ERA5 or GLDAS or NLDAS.

**Reply by Madhuka1990** [reply]
Thanks for the directions. I'll try these

**Comment by Madhuka1990** [comment]
Hi,
I downloaded GLDAS data to run create_forcing.exe to generate the HRLDAS_SETUP_FILE. I extracted the data from raw files using the given code (extract_gldas.py) and as per the variable names in extracted netcdf files, I adjusted the VTable in namelist. 
But when I run the create_forcing.exe, I got error message as "NetCDF data does not match a Vtable entry".
I tried to fix the error, but couldn't.
Could you please help me to solve this issue

**Reply by cenlinhe** [reply]
What part did you change in the VTable in the namelist? Did you use the namelist.input.GLDAS inside the /example/GLDAS/ folder? If so, I do not think you need to adjust the VTable in the namelist. The default one should work.
@CharlesZheZhang Correct me if I am wrong.

**Reply by Madhuka1990** [reply]
First, I ran the create_forcing.exe without changing the VTable in namelist located inside the /example/GLDAS/ folder. Process was stopped and showed an error as "Can't read variable: Tair_f_inst". The variable name in the Tair netcdf files are "Tair_f_inst". Then I changed the VTable by including respective variable names in the forcing data sets.
![Vtable](https://github.com/user-attachments/assets/e729dd25-78d4-4769-8c8a-d7e166bd4445)

**Reply by CharlesZheZhang** [reply]
Hi Madhuka, the Vtable is used to check variable name and assign attributes, matching input and output. Please see this line:
https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/HRLDAS_forcing/create_forcing.F#L2019
The code is actually checking variable names from the array defined here, but not from the Vtable. So, actually you don't need to modify the variable names in the Vtable column, as Cenlin suggested here. 

Best,
Zhe

**Reply by cenlinhe** [reply]
Did you run extract_gldas.py and extract_gldas_init.py before you ran create_forcing.exe?
@CharlesZheZhang We need to create a tutorial doc for creating GLDAS forcing under the /tutorial/folder. I see there are multiple other preprocessing scripts under /examples/GLDAS/ folder. Do we need to run all those before create_forcing.exe or just the two extract_gldas*.py files?

**Reply by CharlesZheZhang** [reply]
The two python codes are for extracting forcing data and initial conditions. 
Other codes, .perl and .f90, are used for spliting U,V wind data and combining precipitation. 
I agree we need to create a jupyter notebook tutorial on processing GLDAS data too. I can do this some time in Nov.

**Reply by CharlesZheZhang** [reply]
See this link for detail:
https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/docs/README.GLDAS#L93

**Reply by Madhuka1990** [reply]
I did run the extract_gldas.py but not the extract_gldas_init.py (as this is optional). I am confusing with the create_forcing.exe as in the VTable the variable names are as T2D, Q2D etc., but when after extracting data from raw forcing data the variable names of the netcdf files are as Tair_f_inst, Qair_f_inst etc. for temperature, humidity respectively. That's why I changed the Vtable to align with data variable names. Otherwise it gives the error "Can't read variable:.."

**Reply by cenlinhe** [reply]
> See this link for detail: https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/docs/README.GLDAS#L93

Thanks, Zhe!

**Reply by cenlinhe** [reply]
> I did run the extract_gldas.py but not the extract_gldas_init.py (as this is optional). I am confusing with the create_forcing.exe as in the VTable the variable names are as T2D, Q2D etc., but when after extracting data from raw forcing data the variable names of the netcdf files are as Tair_f_inst, Qair_f_inst etc. for temperature, humidity respectively. That's why I changed the Vtable to align with data variable names. Otherwise it gives the error "Can't read variable:.."

You can follow the instructions here: https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/docs/README.GLDAS#L42-L176

**Reply by cenlinhe** [reply]
The variable names "Tair_f_inst, Qair_f_inst, etc." are the extracted raw variables from GRIB data files and used by create_forcing.exe. You will see these variable names in create_forcing.f. They are not the output variable names in your forcing files.

**Reply by Madhuka1990** [reply]
Thanks @cenlinhe . I'll follow the instructions and try to sort it out.

**Reply by tslin2** [reply]
there are two discussions for references
https://github.com/NCAR/hrldas/discussions/91
https://github.com/NCAR/hrldas/discussions/103

**Comment by Madhuka1990** [comment]
I followed the steps using global data set as suggested in one of the above discussions. But create_forcing.exe only read the Tair, Qair, Psurf, U, V, LWdown & SWdown files and at the end process was stopped with following error message.
"STOP String is all blanks."
![image](https://github.com/user-attachments/assets/adce779d-e619-4220-a8e5-9e4f00feec54)

Precipitation and initial condition files were not read during the process as per my understanding.
I checked those files as well which are readable and having data.

Could you please let me know how to solve this issue

**Reply by cenlinhe** [reply]
This is weird. @CharlesZheZhang or @tslin2 Could you please take a look? Maybe download a few GLDAS example files and run the create_forcing from your end to see if you have the same issue here. Thanks!

**Reply by Madhuka1990** [reply]
I have a doubt on the create_forcing.F file as it is not reading the Precip files. I used following Makefile to generate that .F file though which name is not similar to name as in discussion in [NCAR/hrldas GitHub](https://github.com/NCAR/hrldas/discussions/91) (create_forcing_netcdf.F)
Could you please whether there is an error in that Makefile?
[makefile.zip](https://github.com/user-attachments/files/17441721/makefile.zip)

**Reply by CharlesZheZhang** [reply]
Hi Madhuka, we used to have 2 .F files (create_forcing.F for grib format forcing and create_forcing_netcdf.F for netcdf forcing). Then we combined these two .F file into one (create_forcing.F) and add "forcing_type" to specify subroutines to read netcdf or grib data. 
As for your question, it seems the code stops after successfully reading SWdown and the next variable is Precf_tavg and it shows "String" is all blank. Did you run combine_precips_netcdf.perl (and the associated compiled .f90 code)? This code will combine the rain and snow data to precipitation and will change the variable names (from Rainf_tavg and Snowf_tavg to Precf_tavg):
https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/HRLDAS_forcing/run/examples/GLDAS/combine_precips_netcdf.f90#L15C1-L17C59
Hope this can help

**Reply by Madhuka1990** [reply]
@CharlesZheZhang, Thanks for the reply.
I ran the combine_precips_netcdf.perl and combined precipitation files were also generated. The details of one of the generated precipitation file is as follows;
![image](https://github.com/user-attachments/assets/e532228a-32ab-423a-a18d-d39d740f27ea).

Is there any way to resolve this

**Reply by tslin2** [reply]
> This is weird. @CharlesZheZhang or @tslin2 Could you please take a look? Maybe download a few GLDAS example files and run the create_forcing from your end to see if you have the same issue here. Thanks!

I am starting to test this; it looks like there are TabError: inconsistent use of tabs and spaces in indentation for python 3 in the extract_gldas.py file

**Comment by Madhuka1990** [comment]
I reran the combine_precips and then ran the create_forcing.exe, but still same error is there.
What can I do to solve this

**Reply by CharlesZheZhang** [reply]
Hi Madhuka, the error seems related to failing to search for the correct variable name associated with precipitation, when looping through all vtable entries (VARSEARCH), so it returns a "blank" string value to "VAR_NAME". 
https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/HRLDAS_forcing/create_forcing.F#L2028 
So it will triggers a STOP when calling VAR_NAME when reading netcdf in later lines. 
https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/HRLDAS_forcing/create_forcing.F#L2051
My suggestion may be adding some write(*,*) functions in the VARSEARCH loop 
"write(*,*) current%label,  vtable(j)%name", so that you know if the current label can match with a vtable entry, hence, passing the correct variable name for precipitation.
Hope this can help.

**Reply by Madhuka1990** [reply]
Hi @CharlesZheZhang, still it is showing same error and no any label relevant to precipitation data are not shown though I added a "write" in VARSEARCH loop. Do I need to update anything in subroutine netcdf_file_search_now in the create_forcing.F (line 1494), because the commands not even check precipitation files in Precip directory.

**Reply by Madhuka1990** [reply]
Hi All,
I was able to resolve the issue. Thanks for your time and helps

**Reply by CharlesZheZhang** [reply]
Hi @Madhuka1990 , great to hear that you are able to resolve this issue. 
Could you also let us know what was the cause and how did you resolve it? So we can make a note and help with future users.
Many thanks! Zhe

**Reply by Madhuka1990** [reply]
Hi @CharlesZheZhang, it was my mistake. Since I didn't use secondary data for SW and precipitation, I commented out all the command lines which processes are relevant to use secondary data, but I was missed out one location (in line 364 of create_forcing.F). After correcting that program was run continuously through each variable.
But again, I have come across another set of issues and concerns as follows
1. "Source model land/sea data not found; assuming a bitmap exists" : Do I need to include separate landmask file or will it be possible to use landmask from geo_em.d01.nc file?
2. Since I'm using 3 hourly data, where can I specify that, as programm is looking for hourly data because increment in "ihour" is 1.
![image](https://github.com/user-attachments/assets/a0dc1d6b-d3cf-4641-81e1-f87f28b36d38)

And the process was stopped again by giving above error.

Can suggest some options to resolve these as well?

Thanks
Madhuka

**Reply by Madhuka1990** [reply]
I have sorted error. Could you guys give some comments on my first 2 concerns

**Reply by CharlesZheZhang** [reply]
Hi @Madhuka1990, for your two comments:
1 source model land/sea mask is needed (assuming you are using the MERRA2 data, correct?), which you can specify in the namelist: LANDSFILE_TEMPLATE = "<DataDir>/<YYYY>/<MM>/<DD>/LANDSEA.<date>". It is needed because the program will need to interpolate from source data (MERRA2, coarse resolution) to destination domain (geo_em.d01.nc, high resolution), to avoid discrepancy of reading data usually occur near coastal line. You may want to look for this land/sea mask data from MERRA2 source.
2 I don't remember it is necessary to specify ihour in the namelist, it is a variable read from the forcing data timestep
L346: call geth_idts(date(1:13), startdate(1:13), ihour)
https://github.com/NCAR/hrldas/blob/7d183b9ba6ff82f746a978fb20bf79e1cb4f1466/hrldas/HRLDAS_forcing/create_forcing.F#L346C6-L346C56

**Comment by Madhuka1990** [comment]
Hi @cenlinhe 
As you mentioned previously, I generated the HRLDAS_setup file by using GLDAS to use with MERRA2 data. When I ran the hrldas.exe with that setup file, I got following error:
![image](https://github.com/user-attachments/assets/1876c392-f936-4c4f-bb57-7e008adc1d53)

I tried to solve this, but couldn't. Could you please help me with this

**Reply by cenlinhe** [reply]
did you modify anything in the parameter table NoahmpTable.TBL? Looks like the table formating has some issues. try re-downloading the table file to your run directory.

**Reply by Madhuka1990** [reply]
I didn't modify the NoahmpTable.TBL in the noahmp/parameters directory, but I have updated the directory path mentioned in the NoahmpTable.TBL and other .TBL files in hrldas/run directory.
I replaced the NoahmpTable.TBL file with the file available in repository and ran the hrldas.exe again. But same error was popping and process was terminated.

**Reply by cenlinhe** [reply]
What is the line 677 of your NoahmpReadTableMod.f90?

**Reply by Madhuka1990** [reply]
Line 677 is " NoahmpIO%ISBARREN_TABLE        = ISBARREN"
![image](https://github.com/user-attachments/assets/7d001b7e-e482-4b6a-8c51-417b83159603)

Do I need to run "make" command within the "run" directory prior to running hrldas.exe?

**Reply by cenlinhe** [reply]
no,  you do not need to run make under "run" directory. Please check your hrldas_setup file to see if you have ISBARREN value defined in the global attribute of this netcdf file. You can use "ncdump -h hrldas_setup_xxx.nc"

**Reply by Madhuka1990** [reply]
I checked the hrldas_setup file and there are no ISBARREN value specified in the global attribute.
How could I do that to solve this issue

**Reply by cenlinhe** [reply]
do you have a geo_em file that was used to create hrldas_setup and forcing files during create_forcing.exe step? If so, you should be able to find all those global attribute in that geo_em file by using ncdump -h geo_em command. And then you can use nco command to copy those global attributes from the geo_em file directly to your hrldas_setup file. (These attributes are supposed to be copied to your hrldas_setup file during the create_forcing.exe step, but I am not sure why they were not copied over). In addition to ISBARREN, please also pay attention to those other global attributes, e.g., ISURBAN, ISICE, etc. those are needed.

**Reply by Madhuka1990** [reply]
I added the ISBARREN to hrldas_setup file.
![image](https://github.com/user-attachments/assets/84d6c8bd-5229-4e1a-b31e-18a7082213de)

But still give same error mentioning that line 677 of file NoahmpReadTableMod.f90 when I ran hrldas.exe.

**Comment by Madhuka1990** [comment]
Since I'm using MERRA2 data where one data file contains 24 hours of data, I separated the file into 24 individual hourly data file by using ncks -O -d Time,$hour,$hour "$INPUT_FILE" "$OUTPUT_FILE" command. Is this approach correct?
Then I ran the hrldas.exe using those hourly data files, but process was stopped stating that "STOP Error: Energy budget problem in NoahMP LSM" . How I correct this error?

**Reply by cenlinhe** [reply]
i have not used this ncks command before, but if you just randomly select a few forcing files to double check if the hourly file has the correct values (i.e., same as those in your daily file). If so, I believe your ncks command should be doing the right thing.
Your error message may be caused by two things: (1) unrealistic values in your forcing file; or (2) your model timestep is too large. You can use a smaller NOAH_TIMESTEP (i.e., use 30-min or 15-min). Also, if your model stopped at the first time step, you may also want to check the initial condition file HRLDAS_setup file to make sure there is no unrealistic value.
